sdk-extensions/tracing-incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/trace/LeakDetectingSpanProcessor.java

```diff
@@ -0,0 +1,153 @@
+    StackTraceElement[] stackTrace = caller.getStackTrace();
+
+    // take off the first 3 stack frames, as they are from the SDK itself.
+    caller.setStackTrace(
+        Arrays.copyOfRange(stackTrace, Math.min(3, stackTrace.length), stackTrace.length));
+
+    pendingSpans.put(span.getSpanContext(), caller);
+  }
+
+  @Override
+  public boolean isStartRequired() {
+    return true;
+  }
+
+  @Override
+  public void onEnd(ReadableSpan span) {
+    pendingSpans.remove(span.getSpanContext()).ended = true;
+  }
+
+  @Override
+  public boolean isEndRequired() {
+    return true;
+  }
+
+  private static class PendingSpans extends WeakConcurrentMap<SpanContext, CallerStackTrace> {
+
+    private final ConcurrentHashMap<WeakKey<SpanContext>, CallerStackTrace> map;
+    private final BiConsumer<String, Throwable> reporter;
+
+    @SuppressWarnings("ThreadPriorityCheck")
+    private static PendingSpans create(BiConsumer<String, Throwable> reporter) {
+      PendingSpans pendingSpans = new PendingSpans(new ConcurrentHashMap<>(), reporter);
+      // Start cleaner thread ourselves to make sure it runs after initializing our fields.
+      Thread thread = new Thread(pendingSpans);
+      thread.setName("weak-ref-cleaner-leakingspandetector");
+      thread.setPriority(Thread.MIN_PRIORITY);
+      thread.setDaemon(true);
+      thread.start();
+      return pendingSpans;
+    }
+
+    private PendingSpans(
+        ConcurrentHashMap<WeakKey<SpanContext>, CallerStackTrace> map,
+        BiConsumer<String, Throwable> reporter) {
+      super(/* cleanerThread= */ false, /* reuseKeys= */ false, map);
+      this.map = map;
+      this.reporter = reporter;
+    }
+
+    // Called by cleaner thread.
+    @Override
+    public void run() {
+      try {
+        while (!Thread.interrupted()) {
+          // call blocks until something is GC'd.
+          Reference<? extends SpanContext> gcdReference = remove();
+          CallerStackTrace caller = map.remove(gcdReference);
+          if (caller != null && !caller.ended) {
+            reporter.accept("Span garbage collected before being ended.", callerError(caller));
+          }
+        }
+      } catch (InterruptedException ignored) {
+        // do nothing
+      }
+    }
+  }
+
+  private static class CallerStackTrace extends Throwable {
+
+    private static final long serialVersionUID = 1234567896L;
+
+    final String threadName = currentThread().getName();
+    final String spanInformation;
+
+    volatile boolean ended;
+
+    CallerStackTrace(ReadableSpan span) {
+      super("Thread [" + currentThread().getName() + "] started span : " + span + " here:");
+      this.spanInformation = span.getName() + " [" + span.getSpanContext() + "]";
+    }
+  }
+
+  private static AssertionError callerError(CallerStackTrace caller) {
+    // Sometimes unit test runners truncate the cause of the exception.
+    // This flattens the exception as the caller of end() isn't important vs the one that leaked
```

Comment:
This comment doesn't apply, `end()` isn't actually called at all - the context detection has the two duties of checking leaked scopes, and checking scopes closed on the wrong thread.

Reply:
will remove; thanks